scene/gui/control.cpp

```diff
@@ -1419,13 +1419,29 @@ void Control::_set_global_position(const Point2 &p_point) {
 
 void Control::set_global_position(const Point2 &p_point, bool p_keep_offsets) {
 	ERR_MAIN_THREAD_GUARD;
+
+	if (p_point == get_global_position()) {
+		return; // Edge case, but avoids calculation.
+	}
+
 	Transform2D inv;
 
 	if (data.parent_canvas_item) {
 		inv = data.parent_canvas_item->get_global_transform().affine_inverse();
 	}
 
-	set_position(inv.xform(p_point), p_keep_offsets);
+	Point2 new_position;
+	if (data.pivot_offset.is_zero_approx() || (data.scale.is_equal_approx(Vector2(1.0, 1.0)) && Math::is_zero_approx(Math::fmod(data.rotation, (real_t)Math_TAU)))) {
+		// The pivot point is at the origin or simply translated.
+		new_position = inv.xform(p_point);
+	} else {
+		inv = Transform2D(0.0, -get_position()) * inv; // Translation transformation in parent space.
+		inv = _get_internal_transform().affine_inverse() * inv; // Internal transformation relative to pivot point.
```

Comment:
This is equivalent to `inv = get_transform().affine_inverse() * inv;` as:
```
   _get_internal_transform().affine_inverse() * Transform2D(0.0f, -get_position())
== _get_internal_transform().affine_inverse() * Transform2D(0.0f, get_position()).affine_inverse()
== (Transform2D(0.0f, get_position()) * _get_internal_transform()).affine_inverse()
== get_transform().affine_inverse()
```
(see #84840)

Further, given `inv` is initially the inverse of the parent global transform:
```
   get_transform().affine_inverse() * parent_global_transform.affine_inverse()
== (parent_global_transform * get_transform()).affine_inverse()
== get_global_transform().affine_inverse()
```

So the whole calculation could be simplified like:
```diff
diff --git a/scene/gui/control.cpp b/scene/gui/control.cpp
index 349102fafb..c352dd4a14 100644
--- a/scene/gui/control.cpp
+++ b/scene/gui/control.cpp
@@ -1419,13 +1419,8 @@ void Control::_set_global_position(const Point2 &p_point) {
 
 void Control::set_global_position(const Point2 &p_point, bool p_keep_offsets) {
 	ERR_MAIN_THREAD_GUARD;
-	Transform2D inv;
-
-	if (data.parent_canvas_item) {
-		inv = data.parent_canvas_item->get_global_transform().affine_inverse();
-	}
-
-	set_position(inv.xform(p_point), p_keep_offsets);
+	Point2 new_position = get_global_transform().affine_inverse().xform(p_point) + get_position();
+	set_position(new_position, p_keep_offsets);
 }
 
 Point2 Control::get_global_position() const {
```

Not sure if it needs a separate case for no pivot offset / (no rotation + unit scale).

Reply:
> Not sure if it needs a separate case for no pivot offset / (no rotation + unit scale).

If using its own `get_global_transform()`, no distinction is needed. The `global_transform` might already be cached.